File: apps/mobile/smarttaxi_app/test/address_resolution_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:smarttaxi_app/features/shared/models.dart';

void main() {
  const address = {'label': 'улица Абая, 14', 'lat': 40.7, 'lng': 68.52};
  test('unresolved server response is blocked even with a convincing label',
      () {
    expect(
        AddressSuggestion.fromJson({...address, 'fallback': true}).isResolved,
        isFalse);
    expect(AddressSuggestion.fromJson(address).isResolved, isTrue);
  });
  test('missing, nonnumeric and out-of-range coordinates cannot be confirmed',
      () {
    for (final bad in [
      null,
      '',
      'no coordinate',
      double.nan,
      double.infinity,
      91
    ]) {
      expect(AddressSuggestion.fromJson({...address, 'lat': bad}).isResolved,
          isFalse,
          reason: '$bad');
    }
    expect(AddressSuggestion.fromJson({...address, 'lng': 181}).isResolved,
        isFalse);
    expect(AddressSuggestion.fromJson({...address, 'lat': '40.7'}).isResolved,
        isTrue);
  });
}

File: apps/web/tools/address-resolution.test.mjs
import assert from 'node:assert/strict';
import test from 'node:test';
import { isResolvedAddressPoint } from '../src/features/client/address-resolution.mjs';

const address = { label: 'улица Абая, 14', lat: 40.7, lng: 68.52 };
test('unresolved response cannot be confirmed even with a readable label', () => {
  assert.equal(isResolvedAddressPoint(address), true);
  assert.equal(isResolvedAddressPoint({ ...address, fallback: true }), false);
});
test('missing and invalid coordinates cannot silently become zero', () => {
  for (const lat of [null, undefined, '', ' ', false, [], {}, NaN, Infinity, 91]) {
    assert.equal(isResolvedAddressPoint({ ...address, lat }), false, String(lat));
  }
  assert.equal(isResolvedAddressPoint({ ...address, lng: 181 }), false);
  assert.equal(isResolvedAddressPoint({ ...address, lat: '40.7' }), true);
  assert.equal(isResolvedAddressPoint({ ...address, lat: 0 }), true);
});
